paasta_tools/check_kubernetes_services_replication.py

```diff
@@ -0,0 +1,164 @@
+from paasta_tools import monitoring_tools
+from paasta_tools.kubernetes_tools import filter_pods_by_service_instance
+from paasta_tools.kubernetes_tools import get_all_nodes
+from paasta_tools.kubernetes_tools import get_all_pods
+from paasta_tools.kubernetes_tools import is_pod_ready
+from paasta_tools.kubernetes_tools import KubeClient
+from paasta_tools.kubernetes_tools import KubernetesDeploymentConfig
+from paasta_tools.long_running_service_tools import get_proxy_port_for_instance
+from paasta_tools.paasta_service_config_loader import PaastaServiceConfigLoader
+from paasta_tools.smartstack_tools import KubeSmartstackReplicationChecker
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import list_services
+from paasta_tools.utils import load_system_paasta_config
+
+
+log = logging.getLogger(__name__)
+
+
+def parse_args():
+    epilog = "PERCENTAGE is an integer value representing the percentage of available to expected instances"
+    parser = argparse.ArgumentParser(epilog=epilog)
+
+    parser.add_argument(
+        '-d', '--soa-dir', dest="soa_dir", metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+    parser.add_argument(
+        '-v', '--verbose', action='store_true',
+        dest="verbose", default=False,
+    )
+    options = parser.parse_args()
+
+    return options
+
+
+def check_healthy_kubernetes_tasks_for_service_instance(
+    instance_config,
+    expected_count,
+    all_pods,
+):
+    si_pods = filter_pods_by_service_instance(
+        pod_list=all_pods,
+        service=instance_config.service,
+        instance=instance_config.instance,
+    )
+    num_healthy_tasks = len([pod for pod in si_pods if is_pod_ready(pod)])
+    log.info(f"Checking {instance_config.service}.{instance_config.instance} in kubernetes as it is not in smartstack")
+    monitoring_tools.send_replication_event_if_under_replication(
+        instance_config=instance_config,
+        expected_count=expected_count,
+        num_available=num_healthy_tasks,
+    )
+
+
+def check_service_replication(
+    instance_config: KubernetesDeploymentConfig,
+    all_pods,
+    smartstack_replication_checker: KubeSmartstackReplicationChecker,
+) -> None:
+    """Checks a service's replication levels based on how the service's replication
+    should be monitored. (smartstack or k8s)
+
+    :param instance_config: an instance of KubernetesDeploymentConfig
+    :param smartstack_replication_checker: an instance of KubeSmartstackReplicationChecker
+    """
+    expected_count = instance_config.get_instances()
+    log.info("Expecting %d total tasks for %s" % (expected_count, instance_config.job_id))
+    proxy_port = get_proxy_port_for_instance(instance_config)
+
+    registrations = instance_config.get_registrations()
+    # if the primary registration does not match the service_instance name then
+    # the best we can do is check marathon for replication (for now).
```

Comment:
s/marathon/kubernetes

Reply:
ps: can we do better than this and figure out what to do with namespaces with multiple instances advertising to it?

Reply:
yeah, I still think we should do that but I guess it's best left to a separate ticket, it is tricky because we can have two instances advertising into the same namespace etc...